tools/wptrunner/wptrunner/environment.py

```diff
@@ -311,14 +311,22 @@ def test_servers(self):
         return failed, pending
 
 
-def wait_for_service(logger, host, port, timeout=60):
+def wait_for_service(logger, host, port, timeout=60, server_process=None):
     """Waits until network service given as a tuple of (host, port) becomes
-    available or the `timeout` duration is reached, at which point
-    ``socket.error`` is raised."""
+    available, `timeout` duration is reached, or the `server_process` exits at
+    which point ``socket.error`` is raised."""
     addr = (host, port)
     logger.debug(f"Trying to connect to {host}:{port}")
     end = time.time() + timeout
     while end > time.time():
+        if server_process is not None and server_process.poll() is not None:
+            returncode = server_process.poll()
+            logger.debug(
+                f"Server process {server_process.pid} exited with "
+                f"{returncode}, giving up trying to connect"
```

Comment:
You could just use `server_process.returncode` here and avoid the additional call to `poll()`? But it doesn't really matter ofc.

Reply:
> You could just use `server_process.returncode` here and avoid the additional call to `poll()`? But it doesn't really matter ofc.

https://searchfox.org/mozilla-central/rev/da1e39f50ef43145623938141c970437b226da9e/testing/mozbase/mozprocess/mozprocess/processhandler.py#909 doesn't actually set `returncode`, only `wait()` does, so `returncode` doesn't necessarily exist on the `ProcessHandler` object yet.